### PR TITLE
Add unit tests for sticky lua module

### PR DESCRIPTION
**Pull request by fmejia97**
**What this PR does / why we need it**:
This PR adds unit tests to `sticky` lua module. We need this because we don't have unit test coverage for this module.

Issue: https://github.com/Shopify/edgescale/issues/435

### CODE REVIEW

File: rootfs/etc/nginx/lua/test/balancer/sticky_test.lua [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+                local default_hash_fn = "md5"
+                local default_hash_implementation = util[default_hash_fn .. "_digest"]
+                assert.equal(sticky_balancer_instance.digest_func, default_hash_implementation)
+            end)
+        end)
+    end)
+
+    describe("balance()", function()
+        setup(function()
+            util.sha1_digest = function(msg) return msg, false end -- Don't hash for simplification
```
Comment (ElvinEfendi) [inline rootfs/etc/nginx/lua/test/balancer/sticky_test.lua]: comment is not adding any value here imo

File: rootfs/etc/nginx/lua/test/balancer/sticky_test.lua [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+            it("should pick an endpoint for the client", function()
+                local cookie = require("resty.cookie")
+                cookie.new = function(self) 
+                    return { 
+                        get = function(self, n) return false end,
+                        set = function(self, n) return true end
+                    }
+                end
+                local sticky_balancer_instance = sticky:new(test_backend)
+                local s = spy.on(sticky_balancer_instance.instance, "next")
```
Comment (ElvinEfendi) [inline rootfs/etc/nginx/lua/test/balancer/sticky_test.lua]: Can we avoid this? This test should not know the implementation details of `balance` function. Why don't you assert on the return value of `:balance`?

File: rootfs/etc/nginx/lua/test/balancer/sticky_test.lua [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+                    local cookie_instance = {
+                        set = function(v) return true, nil end,
+                        get = function(k) end,
+                    }
+                    s = spy.on(cookie_instance, "set")
+                    return cookie_instance, false
+                end
+                local sticky_balancer_instance = sticky:new(get_test_backend())
+                assert.has_no.errors(function() sticky_balancer_instance:balance() end)
+                assert.spy(s).was_called()
```
Comment (ElvinEfendi) [inline rootfs/etc/nginx/lua/test/balancer/sticky_test.lua]: Can you also assert on the parameters of cookie? i.e what key, value, path, domain it is setting.

File: rootfs/etc/nginx/lua/test/balancer/sticky_test.lua [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+            end)
+        end)
+
+        context("when client has a cookie set", function()
```
Comment (ElvinEfendi) [inline rootfs/etc/nginx/lua/test/balancer/sticky_test.lua]: You are missing a crucial test here. We should make sure `:balance` returns the correct endpoint for given cookie value.

File: rootfs/etc/nginx/lua/test/balancer/sticky_test.lua (hunk 2 above)
Comment (fmejia97) [reply rootfs/etc/nginx/lua/test/balancer/sticky_test.lua]: Good point, it currently feels a bit hacky. I like the approach you suggested 👍

**Comment by fmejia97** [comment]
@ElvinEfendi Addressed your comments

File: rootfs/etc/nginx/lua/test/balancer/sticky_test.lua [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+                cookie.new = function(self) 
+                    local return_obj = {
+                        get = function(k) return test_backend_endpoint end,
+                    }
+                    return return_obj, false
+                end
+                local sticky_balancer_instance = sticky:new(test_backend)
+                local ip, port = sticky_balancer_instance:balance()
+                assert.equal(ip .. ":" .. port, test_backend_endpoint)
+            end)
```
Comment (ElvinEfendi) [inline rootfs/etc/nginx/lua/test/balancer/sticky_test.lua]: This test isn't covering the logic properly:
```
> ingress (add-sticky-module-unit-tests)$ git diff
diff --git a/rootfs/etc/nginx/lua/balancer/sticky.lua b/rootfs/etc/nginx/lua/balancer/sticky.lua
index 8af93a98..348cb94b 100644
--- a/rootfs/etc/nginx/lua/balancer/sticky.lua
+++ b/rootfs/etc/nginx/lua/balancer/sticky.lua
@@ -70,7 +70,8 @@ local function sticky_endpoint_string(self)
     set_cookie(self, key)
   end

-  return self.instance:find(key)
+  return key
+  --return self.instance:find(key)
 end

 function _M.balance(self)
> ingress (add-sticky-module-unit-tests)$ make lua-test
Makefile:160: warning: overriding commands for target `verify-all'
Makefile:156: warning: ignoring old commands for target `verify-all'
●●●●●●●●●●●●●●●●●●●●●
21 successes / 0 failures / 0 errors / 0 pending : 0.02952 seconds
```

File: rootfs/etc/nginx/lua/test/balancer/sticky_test.lua [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+                    return return_obj, false
+                end
+                local sticky_balancer_instance = sticky:new(test_backend)
+                assert.has_no.errors(function() sticky_balancer_instance:balance() end)
+                assert.spy(s).was_not_called()
+            end)
+
+            it("should return the correct endpoint for the client", function()
+                local cookie = require("resty.cookie")
+                local s = {}
```
Comment (hnrytrn) [inline rootfs/etc/nginx/lua/test/balancer/sticky_test.lua]: Looks like this variable is unused

**Comment by fmejia97** [comment]
@ElvinEfendi Updated the tests. Removed resty.chash mock since we now have direct access to it.

**Comment by ElvinEfendi** [comment]
this was completed at https://github.com/kubernetes/ingress-nginx/pull/2966

**Pull request closed without merging**